File: geoportal/src/com/esri/gpt/server/csw/client/CswCatalog.java
/* See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * Esri Inc. licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.esri.gpt.server.csw.client;

import java.io.IOException;
import java.io.InputStream;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParserFactory;
import org.apache.commons.httpclient.HttpClient;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

/**
 * Maintains information of CSW Catalog
 * 
 * The catalogs contain all the information like url, profile information
 * credentials and capabilities.
 */
public class CswCatalog implements Comparable {

/** The base url. */
private String baseUrl;

/** The capabilities. */
private CswCatalogCapabilities capabilities = null;

/** The credentials. */
private UsernamePasswordCredentials credentials;

/** The id. */
private String id;

/** The ID counter. */
private static int IDCounter = 1;

/** The is connected. */
private boolean isConnected = false;

/** The locking. */
private boolean locking = false;

/** The name. */
private String name;

/** The profile. */
private CswProfile profile;

/** The url. */
private String url;

/** The request timeout ms. */
private int responseTimeoutMs;

/** The connection timeout ms. */
private int connectionTimeoutMs;

private HttpClient batchHttpClient;



/**
 * Instantiates a new csw catalog.
 */
public CswCatalog() {
}

/**
 * Creates instance of the catalog.
 * 
 * @param url the url
 * @param name the name
 * @param profile the profile
 */
public CswCatalog(String url, String name, CswProfile profile) {
  this.url = url;
  this.id = this.url;
  this.name = name;
  this.profile = profile;
}

/**
  * Gets the underlying Apache HttpClient to be used for batch requests to the
  * same server.
  *
  * @return the batch client
  */
public HttpClient getBatchHttpClient() {
  return this.batchHttpClient;
}

/**
  * Sets the underlying Apache HttpClient to be used for batch requests to the
  * same server.
  *
  * @param batchHttpClient the batch client
  */
public void setBatchHttpClient(HttpClient batchHttpClient) {
  this.batchHttpClient = batchHttpClient;
}

/**
 * Gets the base url.
 * 
 * @return the base url
 */
public String getBaseUrl() {
  return baseUrl;
}

/**
 * Sets the base url.
 * 
 * @param baseUrl the new base url
 */
public void setBaseUrl(String baseUrl) {
  this.baseUrl = baseUrl;
}

/**
 * Gets the connection timeout ms.
 * 
 * @return the connection timeout ms
 */
public int getConnectionTimeoutMs() {
  return connectionTimeoutMs;
}

/**
 * Sets the connection timeout ms.
 * 
 * @param connectionTimeoutMs the new connection timeout ms
 */
public void setConnectionTimeoutMs(int connectionTimeoutMs) {
  this.connectionTimeoutMs = connectionTimeoutMs;
}

/**
 * Gets the request timeout ms.
 * 
 * @return the request timeout ms
 */
public int getResponseTimeoutMs() {
  return responseTimeoutMs;
}

/**
 * Sets the request timeout ms.
 * 
 * @param requestTimeoutMs the new request timeout ms
 */
public void setResponseTimeoutMs(int requestTimeoutMs) {
  this.responseTimeoutMs = requestTimeoutMs;
}

/**
 * Execute GetCapabilities using SAX objects. Send GetCapabilities request,
 * receive the response from a service, and parse the response to get URLs for
 * "GetRecords" and "GetRecordsById".
 * 
 * @return the csw catalog capabilities
 * @throws SAXException the sAX exception
 * @throws IOException Signals that an I/O exception has occurred.
 * @throws ParserConfigurationException the parser configuration exception
 * @return Csw Capabilities object
 */
private CswCatalogCapabilities executeGetCapabilitiesWithSAX()
    throws SAXException, IOException, ParserConfigurationException {
  CswCatalogCapabilities capabilities = new CswCatalogCapabilities();

  CswClient client = new CswClient();
  client.setConnectTimeout(this.getConnectionTimeoutMs());
  client.setReadTimeout(this.getResponseTimeoutMs());
  client.setBatchHttpClient(getBatchHttpClient());
  // Execute submission and parsing into response element
  InputStream responseStream = client.submitHttpRequest("GET", url, "");
 

  SAXParserFactory factory = SAXParserFactory.newInstance();
  factory.setNamespaceAware(true);
  CapabilitiesParse cParse = new CapabilitiesParse(capabilities);
  factory.newSAXParser().parse(new InputSource(responseStream), cParse);

  this.capabilities = capabilities;
  Utils.close(responseStream);
  return capabilities;
}

/**
 * Gets the capabilities.
 * 
 * @return the capabilities
 */
public CswCatalogCapabilities getCapabilities() {
  return this.capabilities;
}

/**
 * Sets the capabilities.
 * 
 * @param capabilities the new capabilities
 */
public void setCapabilities(CswCatalogCapabilities capabilities) {
  this.capabilities = capabilities;
}

/**
 * Gets the credentials.
 * 
 * @return the credentials
 */
public UsernamePasswordCredentials getCredentials() {
  return credentials;
}

/**
 * Sets the credentials.
 * 
 * @param credentials the new credentials
 */
public void setCredentials(UsernamePasswordCredentials credentials) {
  this.credentials = credentials;
}

/**
 * Gets the id.
 * 
 * @return the id
 */
public String getId() {
  return id;
}

/**
 * Sets the id.
 * 
 * @param id the new id
 */
public void setId(String id) {
  this.id = id;
}

/**
 * Gets the name.
 * 
 * @return the name
 */
public String getName() {
  return name;
}

/**
 * Sets the name.
 * 
 * @param name the new name
 */
public void setName(String name) {
  // if there is no input for name, use url as a name
  if (name == null || name.length() == 0) {
    name = url;
  }

  this.name = name;
}

/**
 * Gets the profile.
 * 
 * @return the profile
 */
public CswProfile getProfile() {
  return profile;
}

/**
 * Sets the profile.
 * 
 * @param profile the new profile
 */
public void setProfile(CswProfile profile) {
  this.profile = profile;
}

/**
 * Gets the url.
 * 
 * @return the url
 */
public String getUrl() {
  return url;
}

/**
 * Sets the url.
 * 
 * @param url the new url
 */
public void setUrl(String url) {
  this.url = url;
}

/**
 * To connect to a catalog service. The capabilties details are populated based
 * on the service.
 * 
 * @return true if connection can be made to the csw service
 * @throws ParserConfigurationException the parser configuration exception
 * @throws IOException Signals that an I/O exception has occurred.
 * @throws SAXException the sAX exception
 */
public boolean connect() throws SAXException, IOException,
    ParserConfigurationException {
  // Execute getCapabilites and setup URLs for "GetRecords" and "GetRecordById"
  CswCatalogCapabilities capabilities = executeGetCapabilitiesWithSAX();

  this.isConnected = capabilities.isReady();

  return this.isConnected;
}

/**
 * To test if already connected to a catalog service.
 * 
 * @return true if connection has already been made to the csw service else
 *         false
 */
public boolean IsConnected() {
  return this.isConnected;
}

/**
 * Locking.
 * 
 * @return true, if successful
 */
public boolean Locking() {
  return false;
}


public int compareTo(Object arg0) {
  // TODO: IMPLEMENT THIS LATER
  return 0;
}

}

File: components/desktop/CswClient/trunk/src/CswCatalog.cs
/* See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * Esri Inc. licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using System.Xml.Xsl;
using System.Xml.XPath;


namespace com.esri.gpt.csw { /// <summary>
    /// Maintains information of CSW Catalog
    /// </summary>
    /// <remarks>
    /// The catalogs contain all the information like url, profile information
    /// credentials and capabilities.
    /// </remarks>
    /// 
    public class CswCatalog:IComparable
    {
        private static int IDCounter = 1 ;
        private string id;
        private string name;
        private bool locking;
        private string url;
        private string baseurl;
        private CswProfile profile;
        private CswCatalogCapabilities capabilities;
        private bool isConnect;

        #region "Properties"
        /// <summary>
        /// Name parameter
        /// </summary>
        public string Name {
            set {
                if (value.Length > 0)
                    name = value;
                else
                    name = url;
            }
            get {
                return name;
            }
            
        }
        /// <summary>
        /// URL parameter
        /// </summary>
        public string URL {
            set {
                url = value;
            }
            get {
                return url;
            }
        }
        /// <summary>
        /// BaseURL parameter
        /// </summary>
        public string BaseURL {
            set {
                baseurl = value;
            }
            get {
                return baseurl;
            }
        }
        /// <summary>
        /// Profile parameter
        /// </summary>
        public CswProfile Profile {
            set {
                profile = value;
            }
            get {
                return profile;
            }
        }
        /// <summary>
        /// HashKey parameter
        /// </summary>
        public int HashKey {
            get {
                return url.GetHashCode();
            }
        }

        /// <summary>
        /// ID parameter
        /// </summary>
        public string ID {
            get {
                return id;
            }
        }

        /// <summary>
        /// Locking parameter
        /// </summary>
        public bool Locking {
            set {
                locking = value;
            }
            get {
                return locking;
            }

        }
        /// <summary>
        /// Capabilities parameter
        /// </summary>

        internal CswCatalogCapabilities Capabilities
        {
            get
            {
                return capabilities;
            }
        }


    #endregion                
        # region constructor definition
        /// <summary>
        /// Constructor
        /// </summary>
            public CswCatalog() {
                id = "catalog"+ IDCounter.ToString();
                IDCounter++;
            }
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="surl"></param>
        /// <param name="sname"></param>
        /// <param name="oprofile"></param>
            public CswCatalog(String surl, String sname, CswProfile oprofile) {
                id = "catalog" + IDCounter.ToString();
                IDCounter++;
                URL = surl;
                Profile = oprofile;
                Name = sname;
                Locking = false;
                        
            }

        #endregion
        #region "PrivateFunction"

            /// <summary>
            ///  To retrieve informations about the CSW service.
            /// </summary>
            /// <remarks>
            /// </remarks>
            /// <param name="param1">capabilities baseurl</param>
            /// <returns>Response the get capabilities url</returns>
            private string GetCapabilities(string capabilitiesurl) {
                try {
                    CswClient client = new CswClient();
                    Utils.logger.writeLog("GetCapabilities url : " + capabilitiesurl);
                    string response = client.SubmitHttpRequest("GET", capabilitiesurl, "");
                    Utils.logger.writeLog("GetCapabilities response : " + response);
                    //Console.WriteLine()
                    

                    XmlDocument xmlDoc = new XmlDocument();
                    if (response == null)
                        return null ;
                    xmlDoc.LoadXml(response); 
                    XmlNamespaceManager xmlnsManager = new XmlNamespaceManager(xmlDoc.NameTable);
                    if (this.Profile.CswNamespace.Length <= 0) {
                        this.Profile.CswNamespace = CswProfiles.DEFAULT_CSW_NAMESPACE;
                    }
                    xmlnsManager.AddNamespace("ows", "http://www.opengis.net/ows");
                    xmlnsManager.AddNamespace("csw", this.Profile.CswNamespace);
                    xmlnsManager.AddNamespace("wrs10", "http://www.opengis.net/cat/wrs/1.0");
                    xmlnsManager.AddNamespace("wrs", "http://www.opengis.net/cat/wrs");
                    xmlnsManager.AddNamespace("xlink", "http://www.w3.org/1999/xlink");
                    xmlnsManager.AddNamespace("wcs", "http://www.opengis.net/wcs");
                    if (xmlDoc.SelectSingleNode("/csw:Capabilities|/wrs:Capabilities| /wrs10:Capabilities | /wcs:WCS_Capabilities", xmlnsManager) != null)
                        return response;
                    else
                        return null;
                }
                catch (Exception ex) {
                    Utils.logger.writeLog(ex.StackTrace);
                    throw ex;
                }
            }

            /// <summary>
            ///  To retrieve informations about the CSW service.
            /// </summary>
            /// <remarks>
            /// </remarks>
            /// <param name="param1">capabilitiesxml details </param>
            private bool ParseCapabilities(string capabilitiesxml) {
                try {
                    XmlDocument xmlDoc = new XmlDocument();
                    xmlDoc.LoadXml(capabilitiesxml);
                    XmlNamespaceManager xmlnsManager = new XmlNamespaceManager(xmlDoc.NameTable);
                    if (this.Profile.CswNamespace.Length <= 0) {
                        this.Profile.CswNamespace = CswProfiles.DEFAULT_CSW_NAMESPACE;
                    }
                    xmlnsManager.AddNamespace("ows", "http://www.opengis.net/ows");
                    xmlnsManager.AddNamespace("csw", this.Profile.CswNamespace);
                    xmlnsManager.AddNamespace("wrs", "http://www.opengis.net/cat/wrs");
                    xmlnsManager.AddNamespace("wrs10", "http://www.opengis.net/cat/wrs/1.0");
                    xmlnsManager.AddNamespace("xlink", "http://www.w3.org/1999/xlink");
                    xmlnsManager.AddNamespace("wcs", "http://www.opengis.net/wcs");
                    //setting capabilities
                    if (capabilities == null) capabilities = new CswCatalogCapabilities();

                    if (isWCSService(capabilitiesxml))
                    {
                        capabilities.GetRecordByID_GetURL = xmlDoc.SelectSingleNode("/wcs:WCS_Capabilities/wcs:Capability/wcs:Request/wcs:GetCoverage/wcs:DCPType/wcs:HTTP/wcs:Get/wcs:OnlineResource/@xlink:href", xmlnsManager).Value;

                        if (capabilities.GetRecordByID_GetURL == null || capabilities.GetRecordByID_GetURL.Trim().Length == 0)
                        {
                            Utils.logger.writeLog(Resources.NoValidCoverageUrl);
                            throw new Exception(Resources.NoValidCoverageUrl);
                        }
                    }
                    else
                    {

                        if (this.profile.Name == "INSPIRE CSW 2.0.2 AP ISO")
                        {
                            capabilities.GetRecords_PostURL = xmlDoc.SelectSingleNode("//ows:OperationsMetadata/ows:Operation[@name='GetRecords']/ows:DCP/ows:HTTP/ows:Post[translate(ows:Constraint/ows:Value, 'SOAP', 'soap')='soap']/@xlink:href", xmlnsManager).Value;
                            capabilities.GetRecords_IsSoapEndPoint = true;
                        } else
                            capabilities.GetRecords_PostURL = xmlDoc.SelectSingleNode("//ows:OperationsMetadata/ows:Operation[@name='GetRecords']/ows:DCP/ows:HTTP/ows:Post[translate(ows:Constraint/ows:Value, 'XML', 'xml')='xml']/@xlink:href", xmlnsManager).Value;

                        if (capabilities.GetRecords_PostURL == null || capabilities.GetRecords_PostURL.Trim().Length == 0)
                        {
                            Utils.logger.writeLog(Resources.NoValidPostUrl);
                            throw new Exception(Resources.NoValidPostUrl);
                        }

                        capabilities.GetRecordByID_GetURL = xmlDoc.SelectSingleNode("//ows:OperationsMetadata/ows:Operation[@name='GetRecordById']/ows:DCP/ows:HTTP/ows:Get/@xlink:href", xmlnsManager).Value;
                    }

                    return true;
                }
                catch (Exception e) {
                    Utils.logger.writeLog(e.StackTrace);
                    return false;
                }
                
            }

            private Boolean isWCSService(String capabilitiesXml)
            {
                return capabilitiesXml.IndexOf("WCS_Capabilities") > -1 ? true : false;

            }

        #endregion
        #region "PublicFunction"

            /// <summary>
            ///  To connect to a catalog service.
            ///  The capabilties details are populated based on the service. 
            /// </summary>
            /// <remarks>
            /// </remarks>
            /// <returns>true if connection can be made to the csw service</returns>
            public bool Connect() {
                StringBuilder sb = new StringBuilder();
                sb.AppendLine(DateTime.Now + " Sending GetCapabilities request to URL : " + URL);
                string capabilitesxml = GetCapabilities(URL);            
                if (capabilitesxml != null){
                    sb.AppendLine(DateTime.Now + " GetCapabilitiesResponse xml : " + capabilitesxml);
                    ParseCapabilities(capabilitesxml);
                    sb.AppendLine(DateTime.Now + " Parsed GetCapabilitiesResponse xml...");                
                    isConnect = true;
                    Utils.logger.writeLog(sb.ToString());
                    return true;
                }
                else {
                    sb.AppendLine(DateTime.Now + " Failed to connect to GetCapabilities endpoint.");
                    Utils.logger.writeLog(sb.ToString());
                    isConnect = false;
                    return false;
                }           
            }

            /// <summary>
            ///  To test if already connected to a catalog service.
            /// </summary>
            /// <remarks>
            /// </remarks>
            /// <returns>true if connection has already been made to the csw service else false</returns>
            public bool IsConnected() {        
                   return isConnect;
           
            }
            /// <summary>
            ///  Resets catalog connection
            /// </summary>
            public void resetConnection()
            {
                isConnect = false;

            }
            /// <summary>
            ///  Compares CswCatalog objects by name
            /// </summary>
            public int CompareTo(object obj)
            {
                if (obj is CswCatalog)
                {
                    CswCatalog catalog = (CswCatalog)obj;
                    return this.Name.CompareTo(catalog.Name);
                }
                throw new ArgumentException("object is not a CswCatalog");
            }

        #endregion
    }
}
